File: ios/Runner/BluetoothPlugin.swift
import Flutter
import UIKit
import ExternalAccessory
import CoreBluetooth

public class BluetoothPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var bleManager: BluetoothLeManager?
    private var classicManager: BluetoothClassicManager?
    private var discoveredSink: FlutterEventSink?
    private var pairedSink: FlutterEventSink?
    private var discoveredPendingEvents: [BluetoothEvent] = []
    private var pairedPendingEvents: [BluetoothEvent] = []

    override init() {
        super.init()
        bleManager = BluetoothLeManager { [weak self] event in
            self?.handleBluetoothEvent(event, isPaired: false)
        }
        classicManager = BluetoothClassicManager { [weak self] event in
            self?.handleBluetoothEvent(event, isPaired: true)
        }
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.example/bluetooth", binaryMessenger: registrar.messenger())
        let discoveredChannel = FlutterEventChannel(name: "com.example/bluetooth_discovered_devices", binaryMessenger: registrar.messenger())
        let instance = BluetoothPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        discoveredChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any]
        let isBLE = arguments?["isBLE"] as? Bool ?? false

        switch call.method {
        case "startScanning":
            let timeout = arguments?["timeout"] as? Int
            if isBLE {
                bleManager?.startScanning(timeoutSeconds: timeout, result: result)
            } else {
                classicManager?.startScanning(result: result)
            }

        case "stopScanning":
            if isBLE {
                bleManager?.stopScanning(result: result)
            } else {
                classicManager?.stopScanning(result: result)
            }

        case "pairDevice":
            let deviceName = arguments?["deviceName"] as? String ?? ""
            if isBLE {
                result(FlutterError(code: "NOT_SUPPORTED", message: "Pairing not supported for BLE", details: nil))
            } else {
                classicManager?.pairDevice(deviceName: deviceName, result: result)
            }

        case "getPairedDevices":
            if isBLE {
                result([]) // BLE paired devices not implemented
            } else {
                let pairedDevices = classicManager?.getPairedDevices() ?? []
                result(pairedDevices)
            }

        case "connectToDevice":
            let address = arguments?["address"] as? String ?? ""
            if isBLE {
                bleManager?.connectToDevice(address: address, result: result)
            } else {
                classicManager?.connectToDevice(address: address, result: result)
            }

        case "disconnect":
            if isBLE {
                bleManager?.disconnect(result: result)
            } else {
                classicManager?.disconnect(result: result)
            }

        case "sendData":
            let data = arguments?["data"] as? String
            if isBLE, let data = data {
                bleManager?.sendData(data: data, result: result)
            } else if let data = data {
                classicManager?.sendReceipt(data: data, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Data is required", details: nil))
            }

        case "checkStatus":
            if isBLE {
                bleManager?.checkStatus(result: result)
            } else {
                classicManager?.checkStatus(result: result)
            }

        case "checkBluetoothState":
            let state = bleManager?.getBluetoothState() ?? .unknown
            result(state == .poweredOn ? "poweredOn" : "poweredOff")

        case "checkLocationState":
            result("authorizedWhenInUse") // Simplified for demo

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        let channel = arguments as? String ?? ""
        print("onListen called for \(channel) with arguments: \(String(describing: arguments))")
        if channel == "com.example/bluetooth_discovered_devices" {
            discoveredSink = events
            print("Discovered sink set, processing \(discoveredPendingEvents.count) queued events")
            let eventsToProcess = discoveredPendingEvents
            discoveredPendingEvents.removeAll()
            for event in eventsToProcess {
                if case .discoveredDevice(let device) = event {
                    print("Sending queued discovered device: \(device.name) (\(device.address)), isBLE: \(device.isBle)")
                    events(device.toDictionary())
                }
            }
        } else if channel == "com.example/bluetooth_paired_devices" {
            pairedSink = events
            print("Paired sink set, streaming paired devices and processing \(pairedPendingEvents.count) queued events")
            if let pairedDevices = classicManager?.getPairedDevices() {
                for device in pairedDevices {
                    let pairedDevice = BluetoothEvent.DiscoveredDevice(
                        name: device["name"] as? String ?? "Unknown",
                        address: device["address"] as? String ?? "Unknown",
                        uuids: device["uuids"] as? [String] ?? [],
                        type: "Classic",
                        isBle: false,
                        event: "paired",
                        message: nil
                    )
                    print("Streaming paired device: \(pairedDevice.name) (\(pairedDevice.address))")
                    events(pairedDevice.toDictionary())
                }
            }
            let eventsToProcess = pairedPendingEvents
            pairedPendingEvents.removeAll()
            for event in eventsToProcess {
                if case .pairedDevice(let device) = event {
                    print("Sending queued paired device: \(device.name) (\(device.address)), isBLE: \(device.isBle)")
                    events(device.toDictionary())
                }
            }
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        let channel = arguments as? String ?? ""
        print("onCancel called for \(channel)")
        if channel == "com.example/bluetooth_discovered_devices" {
            discoveredSink = nil
        } else if channel == "com.example/bluetooth_paired_devices" {
            pairedSink = nil
        }
        return nil
    }

    func handleBluetoothEvent(_ event: BluetoothEvent, isPaired: Bool) {
        switch event {
        case .discoveredDevice(let device):

            print("Sending device to discovered sink: \(device.name)")
            if let sink = discoveredSink {
                sink(device.toDictionary())
            } else {
                print("Discovered sink is nil, unable to send device")
                discoveredPendingEvents.append(event)
            }

        case .pairedDevice(let device):

            print("Sending device to paired sink: \(device.name)")
            if let sink = pairedSink {
                sink(device.toDictionary())
            } else {
                print("Discovered sink is nil, unable to send device")
                pairedPendingEvents.append(event)
            }


        case .receivedData(let data):
            print("Received data: \(data.data)")
            let sink = isPaired ? pairedSink : discoveredSink
            if let sink = sink {
                print("Sending received data to \(isPaired ? "paired" : "discovered") sink")
                sink(data.data)
            } else {
                print("\(isPaired ? "Paired" : "Discovered") sink is nil, ignoring received data")
            }
        case .error(let error):
            print("Error occurred: \(error.code), message: \(String(describing: error.message))")
            let sink = isPaired ? pairedSink : discoveredSink
            if let sink = sink {
                print("Sending error to \(isPaired ? "paired" : "discovered") sink")
                sink(FlutterError(code: error.code, message: error.message, details: nil))
            } else {
                print("\(isPaired ? "Paired" : "Discovered") sink is nil, ignoring error")
            }
        default:
            print("Unknown event type")
        }

    }
}

File: android/app/src/main/kotlin/com/example/demo_project_atcom/BluetoothPlugin.kt
package com.example.demo_project_atcom

import android.bluetooth.BluetoothAdapter
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.os.Handler
import android.os.Looper
import android.util.Log

class BluetoothPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var discoveredDevicesChannel: EventChannel
    private lateinit var receivedDataChannel: EventChannel
    private lateinit var pairingStateChannel: EventChannel
    private lateinit var pairedDevicesChannel: EventChannel
    private var discoveredSink: EventChannel.EventSink? = null
    private var receivedDataSink: EventChannel.EventSink? = null
    private var pairingStateSink: EventChannel.EventSink? = null
    private var pairedDevicesSink: EventChannel.EventSink? = null
    private var bluetoothClassicManager: BluetoothClassicManager? = null
    private var bluetoothLeManager: BluetoothLeManager? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val TAG = "BluetoothPlugin"

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.example/bluetooth")
        discoveredDevicesChannel = EventChannel(binding.binaryMessenger, "com.example/bluetooth_discovered_devices")
        receivedDataChannel = EventChannel(binding.binaryMessenger, "com.example/bluetooth_received_data")
        pairingStateChannel = EventChannel(binding.binaryMessenger, "com.example/bluetooth_pairing_state")
        pairedDevicesChannel = EventChannel(binding.binaryMessenger, "com.example/bluetooth_paired_devices")
        methodChannel.setMethodCallHandler(this)
        discoveredDevicesChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                discoveredSink = events
            }
            override fun onCancel(arguments: Any?) {
                discoveredSink = null
            }
        })
        receivedDataChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                receivedDataSink = events
            }
            override fun onCancel(arguments: Any?) {
                receivedDataSink = null
            }
        })
        pairingStateChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                pairingStateSink = events
            }
            override fun onCancel(arguments: Any?) {
                pairingStateSink = null
            }
        })
        pairedDevicesChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                pairedDevicesSink = events
                // Send initial paired devices when stream is listened to
                sendPairedDevices()
            }
            override fun onCancel(arguments: Any?) {
                pairedDevicesSink = null
            }
        })

        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        bluetoothClassicManager = BluetoothClassicManager(context, bluetoothAdapter, ::handleBluetoothEvent)
        bluetoothLeManager = BluetoothLeManager(context, bluetoothAdapter, ::handleBluetoothEvent)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        discoveredDevicesChannel.setStreamHandler(null)
        receivedDataChannel.setStreamHandler(null)
        pairingStateChannel.setStreamHandler(null)
        pairedDevicesChannel.setStreamHandler(null)
        bluetoothClassicManager?.cleanup()
        bluetoothLeManager?.cleanup()
        discoveredSink = null
        receivedDataSink = null
        pairingStateSink = null
        pairedDevicesSink = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startScanning" -> {
                val isBLE = call.argument<Boolean>("isBLE") ?: false
                val timeout = call.argument<Int>("timeout")
                if (isBLE) {
                    bluetoothLeManager?.startScanning(timeout, result)
                } else {
                    bluetoothClassicManager?.startScanning(timeout, result)
                }
            }
            "stopScanning" -> {
                bluetoothClassicManager?.stopScanning(result)
                bluetoothLeManager?.stopScanning(object : Result {
                    override fun success(res: Any?) {
                        result.success(null)
                    }
                    override fun error(code: String, message: String?, details: Any?) {
                        result.error(code, message, details)
                    }
                    override fun notImplemented() {
                        result.notImplemented()
                    }
                })
            }
            "pairDevice" -> {
                val address = call.argument<String>("address")
                if (address != null) {
                    bluetoothClassicManager?.pairDevice(address, result)
                } else {
                    result.error("INVALID_ARGUMENT", "Device address is required", null)
                }
            }
            "connectToDevice" -> {
                val isBLE = call.argument<Boolean>("isBLE") ?: false
                val address = call.argument<String>("address")
                val uuid = call.argument<String>("uuid")
                if (address != null) {
                    if (isBLE) {
                        bluetoothLeManager?.connectToDevice(address, result)
                    } else {
                        bluetoothClassicManager?.connectToDevice(address, uuid, result)
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Device address is required", null)
                }
            }
            "disconnect" -> {
                bluetoothClassicManager?.disconnect(object : Result {
                    override fun success(res: Any?) {
                        bluetoothLeManager?.disconnect(result)
                    }
                    override fun error(code: String, message: String?, details: Any?) {
                        result.error(code, message, details)
                    }
                    override fun notImplemented() {
                        result.notImplemented()
                    }
                })
            }
            "sendData" -> {
                val address = call.argument<String>("address")
                val data = call.argument<String>("data")
                if (address != null && data != null) {
                    bluetoothClassicManager?.sendReceipt(address, data, result)
                    bluetoothLeManager?.sendData(data, object : Result {
                        override fun success(res: Any?) {
                            result.success(null)
                        }
                        override fun error(code: String, message: String?, details: Any?) {
                            result.error(code, message, details)
                        }
                        override fun notImplemented() {
                            result.notImplemented()
                        }
                    })
                } else {
                    result.error("INVALID_ARGUMENT", "Address and data are required", null)
                }
            }
            "checkStatus" -> {
                bluetoothClassicManager?.checkStatus(object : Result {
                    override fun success(res: Any?) {
                        result.success(res)
                    }
                    override fun error(code: String, message: String?, details: Any?) {
                        bluetoothLeManager?.checkStatus(result)
                    }
                    override fun notImplemented() {
                        result.notImplemented()
                    }
                })
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        // Not used directly since we have multiple channels
    }

    override fun onCancel(arguments: Any?) {
        // Not used directly
    }

    private fun handleBluetoothEvent(event: BluetoothEvent) {
        mainHandler.post {
            when (event) {
                is BluetoothEvent.DiscoveredDevice -> {
                    val deviceMap = event.toMap()
                    Log.i(TAG, "Streaming discovered device: $deviceMap")
                    discoveredSink?.success(deviceMap)
                }
                is BluetoothEvent.ReceivedData -> {
                    Log.i(TAG, "Streaming received data: ${event.data}")
                    receivedDataSink?.success(event.data)
                }
                is BluetoothEvent.Error -> {
                    Log.e(TAG, "Streaming error: ${event.code} - ${event.message}")
                    discoveredSink?.success(mapOf("event" to "error", "code" to event.code, "message" to event.message))
                }
                is BluetoothEvent.PairingState -> {
                    val stateMap = event.toMap()
                    Log.i(TAG, "Streaming pairing state: $stateMap")
                    pairingStateSink?.success(stateMap)
                }
            }
        }
    }

    private fun sendPairedDevices() {
        mainHandler.post {
            val classicDevices = bluetoothClassicManager?.getPairedDevices() ?: emptyList()
            val bleDevices = bluetoothLeManager?.getPairedDevices() ?: emptyList()
            val allDevices = (classicDevices + bleDevices).distinctBy { it["address"] }
            allDevices.forEach { device ->
                Log.i(TAG, "Streaming paired device: $device")
                pairedDevicesSink?.success(device)
            }
        }
    }
}